src/blas/backends/cublas/cublas_level1.cpp

```diff
@@ -1010,16 +1097,37 @@ inline sycl::event iamax(const char *func_name, Func func, sycl::queue &queue, i
         onemkl_cublas_host_task(cgh, queue, [=](CublasScopedContextHandler &sc) {
             auto handle = sc.get_handle(queue);
             auto x_ = reinterpret_cast<const cuDataType *>(x);
-            auto int_res_p_ = reinterpret_cast<int *>(int_res_p);
+            if (result_on_device) {
+                cublasSetPointerMode(handle, CUBLAS_POINTER_MODE_DEVICE);
+            }
             cublasStatus_t err;
             // For negative incx, iamax returns 0. This behaviour is similar to that of
             // reference iamax.
-            CUBLAS_ERROR_FUNC_T_SYNC(func_name, func, err, handle, n, x_, incx, int_res_p_);
+            CUBLAS_ERROR_FUNC_T_SYNC(func_name, func, err, handle, n, x_, incx, int_res_p);
+            if (result_on_device) {
+                cublasSetPointerMode(handle, CUBLAS_POINTER_MODE_HOST);
+            }
         });
     });
     done.wait();
-    result[0] = std::max((int64_t)(*int_res_p - 1), int64_t{ 0 });
-    return done;
+    if (result_on_device) {
+        // The following does copy device to host and then host to device
+        // just to adjust to 0-base indexing. This is pretty inefficient, and
+        // should maybe be replaced with a sycl GPU kernel, but it duplicated what
```

Comment:
would prefer avoiding D2H+H2D copies.

Reply:
Buffer API does D2H only I think. If it is a compiler bug, we could open a ticket.

Reply:
From the comment on `iamax` in the buffer API [here](https://github.com/oneapi-src/oneMKL/blob/develop/src/blas/backends/cublas/cublas_level1.cpp#L476):

> This requires to bring the data to host, copy it, and return it back to the device

I am not sure it is a compiler bug. A standalone SYCL kernel runs fine on nvidia hardware, there is some issue with the combination of the SYCL kernels and the cuBLAS calls which may be my mistake. Figuring out exactly what it is may be a significant effort.

In my opinion fixing the bug in #280 and reaching parity with the buffer interface is a sufficient improvement to justify this PR even if performance is not ideal in a case that was not even functional before. If the performance or memory use becomes a problem we can deal with that at a later stage.

Reply:
Agreed. Could you open an issue to this repo to share your findings on adding a SYCL kernel? If/when we work on performance parity with cuBLAS native calls for these APIs, your findings could be useful.

Reply:
Good idea, opened #312